atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java

```diff
@@ -42,45 +47,106 @@
+                new NamedThreadFactory("AsyncInitializer-SerializableTransactionManager", true));
+
+        InitializeCheckingWrapper(SerializableTransactionManager manager,
+                Supplier<Boolean> initializationPrerequisite,
+                Callback callBack) {
+            this.txManager = manager;
             this.initializationPrerequisite = initializationPrerequisite;
+            this.callback = callBack;
+            scheduleInitializationCheckAndCallback();
         }
 
         @Override
         public SerializableTransactionManager delegate() {
+            assertOpen();
             if (!isInitialized()) {
                 throw new NotInitializedException("TransactionManager");
             }
-
-            return manager;
+            return txManager;
         }
 
         @Override
         public boolean isInitialized() {
-            // Note that the PersistentLockService is also initialized asynchronously as part of
-            // TransactionManagers.create; however, this is not required for the TransactionManager to fulfil
-            // requests (note that it is not accessible from any TransactionManager implementation), so we omit
-            // checking here whether it is initialized.
-            return manager.getKeyValueService().isInitialized()
-                    && manager.getTimelockService().isInitialized()
-                    && manager.getTimestampService().isInitialized()
-                    && manager.getCleaner().isInitialized()
-                    && initializationPrerequisite.get();
+            assertOpen();
+            return callbackDone && isInitializedInternal();
         }
 
         @Override
         public LockService getLockService() {
-            return manager.getLockService();
+            assertOpen();
+            return txManager.getLockService();
         }
 
         @Override
         public void registerClosingCallback(Runnable closingCallback) {
-            manager.registerClosingCallback(closingCallback);
+            assertOpen();
+            txManager.registerClosingCallback(closingCallback);
+        }
+
+        @Override
+        public void close() {
+            if (isClosed.compareAndSet(false, true)) {
+                callback.blockUntilSafeToShutdown();
+                txManager.close();
+            }
+        }
+
+        private void assertOpen() {
+            if (isClosed.get()) {
+                if (callbackException.isPresent()) {
+                    throw new IllegalStateException("Operations cannot be performed on closed TransactionManager."
+                            + " Closed due to a callback throw.", callbackException.get());
+                }
+                throw new IllegalStateException("Operations cannot be performed on closed TransactionManager.");
+            }
+        }
+
+        private void scheduleInitializationCheckAndCallback() {
+            executorService.schedule(() -> {
+                if (isClosed.get()) {
```

Comment:
I would prefer we coded defensively and caught exceptions at the top level to make sure the follow up task is always scheduled. Otherwise we run the risk of deadlock.